### ISSUE TITLE
what do I need to do to add full python/Language X support?

### ISSUE THREAD

**Issue by skewballfox** [hediet/vscode-debug-visualizer#63]
hey, noob here.  I've never actually contributed to a repo that wasn't mine.

I wanted to contribute to this project, and start adding full support for other languages(python, java, rust, cpp). I saw that data extractors are necessary to convert the data to a serialized json format. 

Are those visualizers supposed to be in that given language or in javascript?
Also, I want to get familiar with the project structure. What would you consider the entry point to be? how does the flow of execution generally look?

**Comment by hediet** [comment]
Hey thanks so much for your interest! Super cool that you want to add support for Python - that's a language I barely know and already wanted to add.

If you like, we can schedule a meeting to discuss how we can add Python! What do you think?

**Comment by hediet** [comment]
> I saw that data extractors are necessary to convert the data to a serialized json format.

Yes.

> Are those visualizers supposed to be in that given language or in javascript?

They are supposed to be in the given language or in whatever additional language the debugger offers.

> Also, I want to get familiar with the project structure. What would you consider the entry point to be? how does the flow of execution generally look?

Key component is the `EvaluationEngine` and `Evaluator` interface and their implementors.
Probably a `PythonEvaluationEngine` will be required.

The most difficult thing for Python would be:
* Identify which data structures Python offers and what visualizations would suit them best (and which additional visualizations might be cool)
* Think of something how the debugger can inject common data extractors into the python runtime so that the user does not need to install libraries
* Think of something how users can implement their own data extractors (if they want to implement an adapter that can extract json for their own data structures)

**Comment by skewballfox** [comment]
> If you like, we can schedule a meeting to discuss how we can add Python! What do you think?

Yeah, that sounds great. When would work for you?

>They are supposed to be in the given language or in whatever additional language the debugger offers.

I'm guessing that should be found somewhere in the [python debugger used by vscode](https://github.com/microsoft/debugpy)?

> Key component is the `EvaluationEngine` and `Evaluator` interface and their implementors.
> Probably a `PythonEvaluationEngine` will be required

`PythonEvaluationEngine` have to be interfacing with the existing debug session right?

sorry if these questions are kinda basic.

**Comment by hediet** [comment]
> When would work for you?

In the next few days I'm very flexible. What about tomorrow 10am CET [here on jitsi](https://meet.jit.si/MainstreamVulnerabilitiesConfineForever)?

**Comment by hediet** [comment]
> PythonEvaluationEngine have to be interfacing with the existing debug session right?

Kind of, but this is pretty easy - I can help you with all that stuff.

**Comment by skewballfox** [comment]
probably later in the day, I'm in CST which puts that around 3am

**Comment by hediet** [comment]
What about this time range: https://www.worldtimebuddy.com/?qm=1&lid=2892794,6&h=2892794&date=2020-7-21&sln=18-23.5

**Comment by skewballfox** [comment]
yeah, any time in the block would be great

**Comment by hediet** [comment]
So 6pm CET then on jitsi (just click [this](https://meet.jit.si/MainstreamVulnerabilitiesConfineForever) link)? Happy to meet you and to discuss the Python support! I think Python programmers are used to visual tools and might really like this extension.

**Comment by skewballfox** [comment]
6pm CET sounds great

yeah, I've been prepping for technical interviews and doing a lot of algorithm problems so this is definitely going to come in handy.

**Comment by hediet** [comment]
> yeah, I've been prepping for technical interviews and doing a lot of algorithm problems so this is definitely going to come in handy.

No worries, I'm a CS student too and you probably won't need complex algorithms for this project ;)

**Comment by bera5186** [comment]
Cool

**Comment by abhishek-ch** [comment]
May I know what is the status for this ?

**Comment by hediet** [comment]
There still is work to do ;) If you would like to help, please reach out!

**Comment by abhishek-ch** [comment]
love to ,  does the fix needs programming in TS/JS or ...

**Comment by hediet** [comment]
Not necessarily!

**Comment by mvoitko** [comment]
I would like to help

**Comment by skewballfox** [comment]
It needs a bit of both TS and Python. I wrote part of the necessary python code, but didn't make it too far with the necessary TS code

**Comment by hediet** [comment]
That is awesome!

1. I think the first step would be to document which python values should be visualized and how they should be visualized. For now, the "how" should be one of the [existing visualizations](https://hediet.github.io/visualization/). All those existing visualizations can be described with JSON, so the goal would be to collect (python value, JSON document) tuples.

2. Then, we can think of python functions that build such JSON documents from such a python value. For the debug visualizer, such a function is called "Data Extractor". These python functions also need to decide whether they can extract a JSON document for a given value and how good the extracted visualization data is in terms of a numerical priority. Each such data extractor should have a unique id.
3. Given a list of those data extractors, an optional preferred data extractor id and some value, we need some logic to return the best visualization data for that value, considering the preferred data extractor id. Also, all applicable data extractors should be reported (so that the user can select one of them, this is already implemented).
4. Now we need a way to make this data extraction library available to all python scripts. In case of javascript, I just eval' the library into the debuggee.

**Comment by bhavintandel** [comment]
I think if python support is added to this project, it will huge benefit to lot of people. 
Just out of curiosity, as python is known for its dynamic typing, how will we tackle that while creating the visuals?

**Comment by hediet** [comment]
@bhavintandel I don't know, I have close to zero experience with python. I guess there is some kind of instanceof check too. Everything else has to test structural matching.

**Comment by Shivamshaiv** [comment]
I think I have the skillset and am willing to take up responsibility of the Python side of the project. But will need a lot of support from others. As I have severe time constraints.

**Comment by isoletslicer** [comment]
Hey, I'd like to help too, have a medium skillset of python here. I wanna do some side projects while in free time at college/work/research. Still, at my work, i use Python tho.

**Comment by hediet** [comment]
@Shivamshaiv @isoletslicer @mvoitko @abhishek-ch Awesome that you are all interested in contributing 🎉
I'm looking forward to great python support 👍
What do you think of my plan?
I think we can collect these value-visualization tuples independently. And then merge and discuss them. Maybe we can already start with the second step, even though we probably may need to change the code later.

**Comment by abhishek-ch** [comment]
Many more are interested in adding support for #python , this is amazing

**Comment by Rdroshan** [comment]
I find this project interesting. Can I also be a part of the python support for this project?
Want to take it as a side project and learn a lot of new things.

**Comment by hediet** [comment]
@Rdroshan Of course! ;) Please have a look at the steps that I outlined above :)
I suggest to start with the (python-value; visualization-data) tuples as gist file!

**Comment by skewballfox** [comment]
This will be a lot easier to implement now that we have more than a few people wanting to work on this. Here's a gist containing the incomplete python implementation of the supported data types, splitting this into multiple files and turning this into a library should be easy enough. 

As I understand the code(I could be wrong), we still need to call this from a typescript file that implements the PyEvaluationEngine, so that it can pass the output of this program to the visualizer. for reference this is the [evaluation engine for javascript](https://github.com/hediet/vscode-debug-visualizer/blob/master/extension/src/EvaluationWatchService/EvaluationEngine/JsEvaluationEngine.ts).

if my understanding of the code is correct, That should be the only necessary typescript component.

**Comment by GrantorShadow** [comment]
Hey everyone, I've been working with python for quite a while now. Would love to contribute to python, how do I get started?

**Comment by hediet** [comment]
@GrantorShadow awesome! I posted a list of steps further up ;)

**Comment by udaypb** [comment]
> @GrantorShadow Hey , got to see the comment thread today, I have medium to good hands on experience with Python and I would like to contribute!! 
Saw the steps you outlined, would love to hear back from you.

**Comment by aaravchandra02** [comment]
Hi, 

Even I would like to contribute to python implementation. This is a really cool project. I checked the 4 steps that you mentioned above. It's my first open source contribution, some guidance apart from the aforementioned steps will be really helpful.

Thank you and hope to hear back from you soon! 
TC.

**Comment by hediet** [comment]
Hey @all! Did you manage to work on some of the steps I suggested? ;)
@skewballfox already opened a PR in #81 (huge thanks for this!). I'm no python expert, it would be awesome if some of you could have a look at it ;)

**Comment by mvoitko** [comment]
@hediet I have already left some comments

**Comment by skewballfox** [comment]
if you guys want to contribute to the python implementation but don't want to merge the python implementation into the existing plugin until completed, you could do pull request on my branch

**Comment by skewballfox** [comment]
Hey I figured I would post an update here. I talked to @hediet a few days ago about how the library needs to interact with the existing extension. 

If I understood the requirements we need some way to make the library available to the python debugger. He was suggesting code injection. 

@GrantorShadow @mvoitko @Shivamshaiv @abhishek-ch @Rdroshan 
if any of you have experience with code injection,  or have some vscode specific insight that would make the library available without injection, I could definitely use the help.

**Comment by skewballfox** [comment]
if anyone wants to tackle the code injection just do a pull request on my branch, it should be just python. we add the necessary typescript components after we have something that can make the classes/functions available to the python debug session(the attached debuggee)

**Comment by aaravchandra02** [comment]
How do I help with python? Can someone pls guide me!
